frontend/web/components/metadata/AddMetadataToEntity.tsx

```diff
@@ -0,0 +1,251 @@
+  const { data: metadataFieldList, isSuccess: metadataFieldListLoaded } =
+    useGetMetadataFieldListQuery({
+      organisation: organisationId,
+    })
+  const {
+    data: metadataModelFieldList,
+    isSuccess: metadataModelFieldListLoaded,
+  } = useGetMetadataModelFieldListQuery({
+    organisation_id: organisationId,
+  })
+
+  const { data: projectFeatureData, isSuccess: projectFeatureDataLoaded } =
+    useGetProjectFlagQuery(
+      {
+        id: entityId,
+        project: projectId,
+      },
+      { skip: entity !== 'feature' || !entityId },
+    )
+
+  const { data: segmentData, isSuccess: segmentDataLoaded } =
+    useGetSegmentQuery(
+      {
+        id: `${entityId}`,
+        projectId: `${projectId}`,
+      },
+      { skip: entity !== 'segment' || !entityId },
+    )
+
+  const [
+    metadataFieldsAssociatedtoEntity,
+    setMetadataFieldsAssociatedtoEntity,
+  ] = useState<CustomMetadataField[]>()
+
+  useEffect(() => {
+    getMetadata?.(metadataFieldsAssociatedtoEntity!)
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [metadataFieldsAssociatedtoEntity])
+
+  useEffect(() => {
+    if (
+      metadataFieldList &&
+      metadataFieldListLoaded &&
+      metadataModelFieldList &&
+      metadataModelFieldListLoaded
+    ) {
+      // Filter metadata fields based on the provided content type
+      const metadataForContentType = metadataFieldList.results
+        // Filter metadata fields that have corresponding entries in the metadata model field list
+        .filter((meta) => {
+          return metadataModelFieldList.results.some((item) => {
+            return (
+              item.field === meta.id && item.content_type === entityContentType
+            )
+          })
+        })
+        // Map each filtered metadata field to include additional information from the metadata model field list
+        .map((meta) => {
+          // Find the matching item in the metadata model field list
+          const matchingItem = metadataModelFieldList.results.find((item) => {
+            return (
+              item.field === meta.id && item.content_type === entityContentType
+            )
+          })
+          // Determine if isRequiredFor should be true or false based on is_required_for array
+          const isRequiredFor =
+            matchingItem &&
+            matchingItem.is_required_for &&
+            matchingItem.is_required_for.length > 0
+
+          // Return the metadata field with additional metadata model field information including isRequiredFor
+          return {
+            ...meta,
+            isRequiredFor: isRequiredFor || false,
+            metadataModelFieldId: matchingItem ? matchingItem.id : null,
+          }
+        })
+      if (projectFeatureData?.metadata && projectFeatureDataLoaded) {
+        const mergeLists = (
+          metadata: ProjectFlag['metadata'],
+          metadataField: CustomMetadataField[],
+        ) => {
+          const map = new Map(
```

Comment:
Im struggling to understand this part, what's actually happening here ?

Reply:
I added comments to explain the functioning of this function